main.go

```diff
@@ -185,42 +211,78 @@ func main() {
 		r.Use(middleware.RealIP)
 		r.Use(middleware.Recoverer)
 		r.Use(middleware.StripSlashes)
-		r.Use(middleware.Timeout(2 * time.Minute)) // best set per handler
+		r.Use(middleware.Timeout(middlewareTimeout)) // best set per handler
 		r.Use(server.Logger(logger))
 
 		ins := server.NewInstrumentationMiddleware(reg)
 
-		r.Mount("/",
-			metricslegacy.NewHandler(
-				cfg.metrics.readEndpoint,
-				metricslegacy.Logger(logger),
-				metricslegacy.Registry(reg),
-				metricslegacy.HandlerInstrumenter(ins),
-			))
-
-		r.Mount("/api/metrics/v1",
-			http.StripPrefix("/api/metrics/v1",
-				metricsv1.NewHandler(
-					cfg.metrics.readEndpoint,
-					cfg.metrics.writeEndpoint,
-					metricsv1.Logger(logger),
-					metricsv1.Registry(reg),
-					metricsv1.HandlerInstrumenter(ins),
-				),
-			),
-		)
+		r.Group(func(r chi.Router) {
+			r.Use(authentication.WithTenant)
+
+			tenantIDs := make(map[string]string)
+			var oidcs []authentication.OIDCConfig
+			for _, t := range tenantsCfg.Tenants {
+				level.Info(logger).Log("msg", "adding a tenant", "tenant", t.Name)
+				tenantIDs[t.Name] = t.ID
+				oidcs = append(oidcs, authentication.OIDCConfig{
+					Tenant:       t.Name,
+					ClientID:     t.OIDC.ClientID,
+					ClientSecret: t.OIDC.ClientSecret,
+					IssuerURL:    t.OIDC.IssuerURL,
+					RedirectURL:  t.OIDC.RedirectURL,
+				})
+			}
 
-		r.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-			r.URL.Path = "/api/metrics/v1/graph"
-			http.Redirect(w, r, r.URL.String(), http.StatusMovedPermanently)
+			tenantHeader := authentication.NewTenantHeaderMiddleware(cfg.metrics.tenantHeader, tenantIDs)
```

Comment:
I thinking inlining this a few lines further down would be nicer for scoping. oidcMiddleware is an exception because it also creates the handler.